qutip/solver/heom/bofin_fit.py

```diff
@@ -0,0 +1,1095 @@
+
+    def set_spectral_density(self, w, J):
+        """
+        Sets the spectral density to be fitted. It may be provided either as an
+        array of function values or as a python function. For internal reasons,
+        it will then be interpolated or discretized as necessary.
+        """
+
+        if callable(J):
+            self._w = w
+            self._J_array = J(w)
+            self._J_fun = J
+        else:
+            self._w = w
+            self._J_array = J
+            self._J_fun = InterpolatedUnivariateSpline(w, J)
+
+    @staticmethod
+    def _meier_tannor_SD(w, a, b, c):
+        r"""
+        Underdamped spectral density used for fitting in Meier-Tannor form
+        (see Eq. 38 in the BoFiN paper, DOI: 10.1103/PhysRevResearch.5.013181)
+        or the get_fit method.
+
+        Parameters
+        ----------
+        w : np.array
+            The frequency of the spectral density
+        a : np.array
+            Array of coupling constants
+        b : np.array
+            Array of cutoff parameters
+        c : np.array
+            Array of resonant frequencies
+        """
+
+        return sum((2 * ai * bi * w
+                    / ((w + ci) ** 2 + bi ** 2)
+                    / ((w - ci) ** 2 + bi ** 2))
+                   for ai, bi, ci in zip(a, b, c))
+
+    def get_fit(
+        self,
+        N=None,
+        Nk=1,
+        final_rmse=5e-6,
+        lower=None,
+        upper=None,
+        sigma=None,
+        guesses=None,
+    ):
+        r"""
+        Provides a fit to the spectral density with N underdamped oscillator
+        baths. N can be determined automatically based on reducing the
+        normalized root mean squared error below a certain threshold.
+
+        Parameters
+        ----------
+        N : optional, int
+            Number of underdamped oscillators to use.
+            If set to None, it is determined automatically.
+        Nk : optional, int
+            Number of exponential terms used to approximate the bath
+            correlation functions, per bath. Defaults to 1.
+        final_rmse : float
+            Desired normalized root mean squared error. Defaults to
+            $5\times10^{-6}$
+        lower : list
+            lower bounds on the parameters for the fit. A list of size 3,
+            each value represents the lower bound for each parameter.
+            The order of the parameters is the same as for the function to be
+            fitted.
+
+            .. math::
+                J(\omega) = \sum_{i=1}^{k} \frac{2 a_{i} b_{i} \omega
+                }{\left(\left( \omega + c_{i}\right)^{2} + b_{i}^{2}\right)
+                \left(\left( \omega - c_{i}\right)^{2} + b_{i}^{2} \right)}
+
+            The first term describes the coupling,
+            the second the cutoff frequency,and the last one the central
+            frequency. The lower bounds are considered to be the same for all
+            N modes. for example
```

Comment:
(Just to make this a bit more concise)

```suggestion
            Lower bounds on the parameters for the fit. A list of size 3,
            containing the lower bounds for :math:`a_i` (coupling constants),
            :math:`b_i` (cutoff frequencies) and :math:`c_i`
            (resonant frequencies) in the following fit function:

            .. math::
                J(\omega) = \sum_{i=1}^{k} \frac{2 a_{i} b_{i} \omega
                }{\left(\left( \omega + c_{i}\right)^{2} + b_{i}^{2}\right)
                \left(\left( \omega - c_{i}\right)^{2} + b_{i}^{2} \right)}

            The lower bounds are considered to be the same for all N modes.
            For example,
```